rust: Close() doesn't reliably cancel listeners, causing panics

Such as:
```
panic: Log in goroutine after TestMultiprocessInitialE2EESyncDoesntDropDeviceListUpdates has completed: [[@user-15-alice:hs1](https://matrix.to/#/@user-15-alice:hs1)]TimelineDiff change: &{ID:$sqDro8apgvTq4Gi8G-pjchrhLyQLvxC4h7jKNOjpfv8 Text:pre message Sender:[@user-16-bob:hs1](https://matrix.to/#/@user-16-bob:hs1) Targe
```
..despite the test calling `.Close()` on the client. It looks like this is a race condition as the logs show:
```
2024-10-16T12:04:19.458875Z  INFO TestMultiprocessInitialE2EESyncDoesntDropDeviceListUpdates: [@user-15-alice:hs1](rust) Close | rust.go:0
2024-10-16T12:04:19.458912Z DEBUG log: cancel    
2024-10-16T12:04:19.458925Z DEBUG matrix_sdk_ffi::task_handle: Cancelling the task handle | bindings/matrix-sdk-ffi/src/task_handle.rs:24
2024-10-16T12:04:19.458956Z DEBUG matrix_sdk_ffi::task_handle: Cancelling the task handle | bindings/matrix-sdk-ffi/src/task_handle.rs:24
2024-10-16T12:04:19.458655Z DEBUG matrix_sdk::sliding_sync: Done handling response | crates/matrix-sdk/src/sliding_sync/mod.rs:679 | spans: sync_once{pos="29/s159_1_0_1_8_1_76_85_0_1"}
```
where subsequent lines then clearly show the timeline listener being invoked, which does `t.Logf` which causes the panic.

Instead, we should latch `closed = true` when closed is called then immediately return on all callbacks `if closed`. It's a bit crap though.

--- a/internal/api/rust/rust.go
+++ b/internal/api/rust/rust.go
@@ -67,6 +67,7 @@ type RustClient struct {
 	userID                string
 	persistentStoragePath string
 	opts                  api.ClientCreationOpts
+	closed                *atomic.Bool
 
 	// for push notification tests (single/multi-process)
 	notifClient *matrix_sdk_ffi.NotificationClient
@@ -103,6 +104,7 @@ func NewRustClient(t ct.TestLike, opts api.ClientCreationOpts) (api.Client, erro
 		roomsMu:               &sync.RWMutex{},
 		opts:                  opts,
 		persistentStoragePath: "./rust_storage/" + username,
+		closed:                &atomic.Bool{},
 	}
 	if opts.AccessToken != "" { // restore the session
 		session := matrix_sdk_ffi.Session{
@@ -295,6 +297,7 @@ func (c *RustClient) ForceClose(t ct.TestLike) {
 
 func (c *RustClient) Close(t ct.TestLike) {
 	t.Helper()
+	c.closed.Store(true)
 	c.roomsMu.Lock()
 	for _, rri := range c.rooms {
 		if rri.stream != nil {
@@ -668,7 +671,9 @@ func (c *RustClient) findRoom(t ct.TestLike, roomID string) *matrix_sdk_ffi.Room
 func (c *RustClient) Logf(t ct.TestLike, format string, args ...interface{}) {
 	t.Helper()
 	c.logToFile(t, format, args...)
-	t.Logf(format, args...)
+	if !c.closed.Load() {
+		t.Logf(format, args...)
+	}
 }
 
 func (c *RustClient) logToFile(t ct.TestLike, format string, args ...interface{}) {

